src/Compilers/Core/Portable/MetadataReference/ModuleMetadata.cs

```diff
@@ -22,58 +22,58 @@ public sealed partial class ModuleMetadata : Metadata
         private readonly PEModule _module;
 
         /// <summary>
-        /// Optional data that should be kept alive as long as this <see cref="ModuleMetadata"/> is alive.  This can be
-        /// useful, for example, if there is backing memory that the metadata depends on that should be kept rooted so it
-        /// doesn't get garbage collected.
+        /// Optional action to invoke when this metadata is disposed.
         /// </summary>
-        private readonly IDisposable? _owner;
+        private readonly Action? _onDispose;
 
         /// <summary>
-        /// Whether or not <see cref="_owner"/> should be <see cref="IDisposable.Dispose"/>'d when this object is
-        /// Disposed.  Is controlled by the <c>leaveOpen</c> flag in <see cref="CreateFromStream(Stream, bool)"/>, or
-        /// the <see cref="PEStreamOptions.LeaveOpen"/> flag in <see cref="CreateFromStream(Stream, PEStreamOptions)"/>.
+        /// Whether or not to call _onDispose when this is disposed.  The reason that this is controlled by a flag is
+        /// that if we make a copy of this modulemetadata, we still want to keep the Action alive in it (as it may be
+        /// keeping alive underlying object this metadata needs).  However, the copy of the metadata should not call the
+        /// action as only the <see cref="Metadata.IsImageOwner"/> metadata is responsible for that.
```

Comment:
This seems dangerous to me. Who says that the copy has a shorter lifetime than the original? Feels like we need some sort of reference count wrapper that is decremented on dispose of the containing reference.

Reply:
i agree ti's dangerous.  BUt it's already the established semantics of ModuleMetadta.  e.g. *prior* to this change if you did this:

```c#
ModuleMetadata mm = ... 
mmCopy = mm.Copy();
mm.Dispose();
```

then `mmCopy` was now in an 'no longer usable' state (same with any other copies created).  I'm genuinely surprised we went with those semantics, but it's been the way of this type since the beginning, so it's just part of the nature of these and hte entire stack is written to be ok with that.

I would totally be ok with moving to a refcounted system.  And, importantly, it would not change this new API at all.  All it would do would be to allow us to remove those comments and make things safer.  As such, i would prefer it be a separate orthogonal API :)

Reply:
> I'm genuinely surprised we went with those semantics, 

For performance reasons. Entirely safe semantics would be slow since every read from the metadata would need to check whether the memory is disposed or not before it reads and do these two operations atomically.

Reply:
> Who says that the copy has a shorter lifetime than the original?

It's responsibility of the owner of the metadata to guarantee that.

Reply:
I guess making the copies ref-counted would add some value. It wouldn't make it entirely safe but perhaps safer.